source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
+    // equivalent functionality (from the user's perspective) as a real consumer
+    // does nothing.
   }
-}
-
-static bool IsCapabilityOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpCapability;
-}
-
-static bool IsExtensionOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpExtension;
-}
-
-static bool IsExtInstImportOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpExtInstImport;
-}
+  ~Impl() { spvContextDestroy(context); }
 
-static bool IsEntryPointOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpEntryPoint;
-}
+  spv_context context;  // C interface context object.
+};
 
-static bool IsExecutionModeOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpExecutionMode;
-}
+SpirvLinker::SpirvLinker(spv_target_env env) : impl_(new Impl(env)) {}
 
-static bool IsDebugAOpcode(spv::Op opcode) {
-  switch (opcode) {
-  case spv::Op::OpString:
-  case spv::Op::OpSourceExtension:
-  case spv::Op::OpSource:
-  case spv::Op::OpSourceContinued:
-    return true;
-  default:
-    return false;
-  }
-}
+SpirvLinker::~SpirvLinker() {}
 
-static bool IsDebugBOpcode(spv::Op opcode) {
-  return opcode == spv::Op::OpName || opcode == spv::Op::OpMemberName;
+void SpirvLinker::SetMessageConsumer(MessageConsumer consumer) {
+  SetContextMessageConsumer(impl_->context, std::move(consumer));
 }
 
-static bool IsTypeOrConstantOpcode(spv::Op opcode) {
-  switch (opcode) {
-  case spv::Op::OpLine:
-  case spv::Op::OpTypeVoid:
-  case spv::Op::OpTypeBool:
-  case spv::Op::OpTypeInt:
-  case spv::Op::OpTypeFloat:
-  case spv::Op::OpTypeVector:
-  case spv::Op::OpTypeMatrix:
-  case spv::Op::OpTypeImage:
-  case spv::Op::OpTypeSampler:
-  case spv::Op::OpTypeSampledImage:
-  case spv::Op::OpTypeArray:
-  case spv::Op::OpTypeRuntimeArray:
-  case spv::Op::OpTypeStruct:
-  case spv::Op::OpTypeOpaque:
-  case spv::Op::OpTypePointer:
-  case spv::Op::OpTypeFunction:
-  case spv::Op::OpTypeEvent:
-  case spv::Op::OpTypeDeviceEvent:
-  case spv::Op::OpTypeReserveId:
-  case spv::Op::OpTypeQueue:
-  case spv::Op::OpTypePipe:
-  case spv::Op::OpTypeForwardPointer:
-  case spv::Op::OpConstantTrue:
-  case spv::Op::OpConstantFalse:
-  case spv::Op::OpConstant:
-  case spv::Op::OpConstantComposite:
-  case spv::Op::OpConstantSampler:
-  case spv::Op::OpConstantNull:
-  case spv::Op::OpSpecConstantTrue:
-  case spv::Op::OpSpecConstantFalse:
-  case spv::Op::OpSpecConstant:
-  case spv::Op::OpSpecConstantComposite:
-  case spv::Op::OpSpecConstantOp:
-  case spv::Op::OpVariable:
-  case spv::Op::OpUndef:
-    return true;
-  default:
-    return false;
+bool SpirvLinker::Link(const std::vector<std::vector<uint32_t>>& binaries,
+                       std::vector<uint32_t>& linked_binary,
+                       uint32_t options) const {
+  auto linkedModule = MakeUnique<Module>();
+
+  std::vector<std::unique_ptr<Module>> modules;
+  modules.reserve(binaries.size());
+
+  uint32_t id_bound = 0u;
```

Comment:
I added some comments in the function to better mark the different steps, as well as turn some steps into a separate function.